src/data/file_preprocessing/echo_files_parser.py

```diff
@@ -0,0 +1,115 @@
+from __future__ import annotations
+import pandas as pd
+
+class EchoFilesParser:
+    def __init__(
+        self,
+        echo_files: list[str],
+    ):
+        """
+        :param echo_files: list of echo file names
+        """
+        self.echo_files = echo_files
+        self.echo_df = pd.DataFrame()
+        self.exceptions_df = pd.DataFrame()
+
+    def find_marker_rows(self, file:str, markers:tuple[str]) -> list[int]:
+        """
+        Finds the row numbers of marker lines in a file.
+
+        :param file: file to search
+        :param markers: markers to search for
+        """
+        with open(file) as f:
+            markers_rows = list()
+            for i, line in enumerate(f):
+                if line.strip() in markers:
+                    markers_rows.append(i)
+                if len(markers_rows) == len(markers):
+                    return markers_rows
+        if len(markers_rows) == 0:
+            raise ValueError('No marker found in file.')
+        return markers_rows
+
+    def parse_files(self) -> EchoFilesParser:
+        """
+        Preprocesses csv echo files, splits regular records from exceptions.
+        """
+        if not(all(file.endswith('.csv') for file in self.echo_files)):
+            raise ValueError(f"Expected files to be of '*.csv' type, provided: {self.echo_files}")
+
+        for filename in self.echo_files: 
+            markers = self.find_marker_rows(filename, ('[EXCEPTIONS]', '[DETAILS]'))
+
+            if len(markers) == 2:
+                exceptions_line, details_line = markers
+                exceptions_df = pd.read_csv(filename, skiprows=exceptions_line+1, nrows=(details_line-1)-exceptions_line-2)
+                echo_df = pd.read_csv(filename, skiprows=details_line+1)
+            else:
+                exceptions_df = pd.DataFrame()
+                echo_df = pd.read_csv(filename, skiprows=markers[0]+1)
+
+            echo_df = echo_df[~echo_df[echo_df.columns[0]].str.lower().str.startswith('instrument')]
+
+            self.exceptions_df = exceptions_df if self.exceptions_df.empty else pd.concat([self.exceptions_df, exceptions_df], ignore_index=True)
+            self.echo_df = echo_df if self.echo_df.empty else pd.concat([self.echo_df, echo_df], ignore_index=True)
+
+        return self
+    
+    def link_bmg_files(self, 
```

Comment:
mutable types (in this case lists) can lead to some nasty bugs if provided as default parameters (since they are the same for all created instances, mutating one will change behaviour of other instances)
you could just omit providing the defaults or create a different way of passing these params (e.g. aggregating them to a common dataclass)

+`echo_keys` param is missing description in the docstring

Reply:
I see, for now I've changed them to tuples